Sources/MapboxNavigation/Array.swift

```diff
@@ -56,48 +56,48 @@ extension Array where Iterator.Element == [CLLocationCoordinate2D]? {
 extension Array where Iterator.Element == CLLocationCoordinate2D {
 
     /**
-     Returns an array of congestion segments by associating the given congestion levels with the coordinates of the respective line segments that they apply to.
+     Returns a dictionary of the fraction of the route leg distance where the congestion level changes and the cooresponding congestion level after the fraction.
      
-     This method coalesces consecutive line segments that have the same congestion level.
-     
-     For each item in the `CongestionSegment` collection a `CongestionLevel` substitution will take place that has a streets road class contained in the `roadClassesWithOverriddenCongestionLevels` collection.
+     For each item in the dictionary,  a `CongestionLevel` substitution will take place that has a streets road class contained in the `roadClassesWithOverriddenCongestionLevels` collection.
      For each of these items the `CongestionLevel` for `.unknown` traffic congestion will be replaced with the `.low` traffic congestion.
      
      - parameter congestionLevels: The congestion levels along a leg. There should be one fewer congestion levels than coordinates.
      - parameter streetsRoadClasses: A collection of streets road classes for each geometry index in `Intersection`. There should be the same amount of `streetsRoadClasses` and `congestions`.
      - parameter roadClassesWithOverriddenCongestionLevels: Streets road classes for which a `CongestionLevel` substitution should occur.
-     - returns: A list of `CongestionSegment` tuples with coordinate and congestion level.
+     - returns: A dictionary of the fraction of the congestion level change start point in route leg, and the congestion level after that.
      */
     func combined(_ congestionLevels: [CongestionLevel],
```

Comment:
Looks like refactored method is the source of performance loss. For example it takes around 10 seconds:
```
Started: 2021-08-03 18:19:31.676258-0700
Finished: 2021-08-03 18:19:41.667548-0700
```
to request the route from San Francisco to Seattle.

Reply:
Yes, you're right, the performance loss in the above is caused by the calculation of the distance. In the previous method, we only calculate the distance of each color slice once in the `func routeLineGradient(_:fractionTraveled:isMain:)` , the `func combined(_:streetsRoadClasses:roadClassesWithOverriddenCongestionLevels:)` only store the coordinates slice for each color.

This pr calculate the distance of each color slice from the start of the leg once in the `func combined`, and also the total leg distance once for the fraction. In the `func routeLineGradient`, it re-calculate the leg distance once. The extra calculation of distance based on the `LineString` should be the cause. Thanks for the suggestion, I'll improve this one.

Reply:
Thanks for the suggestion, I'll removed the line gradients stops generating part. And only keeps the gradient stops updating part, with tests added for empty stops, single unknown stops, and different `fractionTraveled`.